pandas/core/frame.py

```diff
@@ -8827,17 +8881,55 @@ def idxmax(self, axis=0, skipna=True) -> Series:
         Wheat Products     consumption
         Beef              co2_emissions
         dtype: object
+
+        >>> df2 = pd.DataFrame({'a' : [2,4,4,2],
+        ...                    'b' : [4,2,2,4],
+        ...                    'c' : [4,2,2,4]},
+        ...                    index=['A', 'B', 'C', 'D'])
+
+        >>> df2.idxmax(keep='all')
+        a    [B, C]
+        b    [A, D]
+        c    [A, D]
+        dtype: object
+
+        >>> df2.idxmax(keep='last')
+        a    C
+        b    D
+        c    D
+        dtype: object
         """
         axis = self._get_axis_number(axis)
-        indices = nanops.nanargmax(self.values, axis=axis, skipna=skipna)
+        if keep == "last":
+            if axis == 0:
+                indices = nanops.nanargmax(self.values[::-1], axis=axis, skipna=skipna)
+            else:
+                indices = nanops.nanargmax(
+                    self[self.columns[::-1]].values, axis=axis, skipna=skipna
+                )
+            indices = (self.shape[axis] - 1) - indices
+        else:
+            indices = nanops.nanargmax(self.values, axis=axis, skipna=skipna)
 
         # indices will always be np.ndarray since axis is not None and
         # values is a 2d array for DataFrame
         # error: Item "int" of "Union[int, Any]" has no attribute "__iter__"
         assert isinstance(indices, np.ndarray)  # for mypy
 
         index = self._get_axis(axis)
-        result = [index[i] if i >= 0 else np.nan for i in indices]
+        result = []
+        if keep == "all":
+            if axis == 0:
+                for i in range(0, self.shape[1]):
+                    result.append(
```

Comment:
use a list comprehension, this also is terribly inefficient

Reply:
In addition to using a list comprehension, do you have any other suggestions on how this can be improved?